### PR TITLE
Mako Batch/Interactive Templates & Build Fixes

### LINKED ISSUE #292: CI doesn't ensure tests/ is committed
If your run `./mfc.sh test` on `master`, you will observe a `git diff` in case files for some tests. Per @sbryngelson's request for having test case files kept in source control, we should ensure through CI that these files are committed.

The diff currently includes some Phase Change tests (@JRChreim). A few other cases also have diffs but they still describe the same cases - physically.

### PR DESCRIPTION
**Pull request by henryleberre**
The last PR before the Benchmarking one.

Closes #292, #240, and #287.

### CODE REVIEW

File: docs/documentation/running.md [hunk 1]
```diff
@@ -2,7 +2,31 @@
...
+> Running simulations locally should work out of the box. On supported clusters,
+> you can append `-c <computer name>` on the command line to instruct the MFC toolchain
+> to make use of the template file `toolchain/templates/<computer name>.mako`. You can
+> browse that directory and contribute your own files. Since systems and their schedulers
+> do not have a standardized syntax to request certain resources, MFC can only provide
+> support for a restricted subset of common or user-contributed configuration options.
+>
+> Adding a new template file or modifying an existing one will most likely be required if:
+> - You are on a cluster that does not have a template yet.
+> - Your cluster is configured with SLURM and but fails when interactive jobs are
```
Comment (sbryngelson) [inline docs/documentation/running.md]: typo 'and but'

File: misc/run-phoenix-release-cpu.sh [hunk 2]
```diff
@@ -12,5 +12,4 @@ cd "$SLURM_SUBMIT_DIR"
 echo "Running in $(pwd):"
 
 . ./mfc.sh load -c p -m gpu
-./mfc.sh test -j $(nproc) -b mpirun -a
-
+./mfc.sh test -j $(nproc) -a -- -b mpirun
```
Comment (sbryngelson) [inline misc/run-phoenix-release-cpu.sh]: can you comment on what `--` does?

File: toolchain/mfc/args.py [hunk 3]
```diff
@@ -98,29 +94,31 @@ def add_common_arguments(p, mask = None):
...
-    run.add_argument(      "--dry-run",        action="store_true",                       default=False,      help="(Batch) Run without submitting batch file.")
-    run.add_argument("--case-optimization",    action="store_true",                       default=False,      help="(GPU Optimization) Compile MFC targets with some case parameters hard-coded.")
-    run.add_argument(      "--no-build",       action="store_true",                       default=False,      help="(Testing) Do not rebuild MFC.")
-    run.add_argument("--wait",                 action="store_true",                       default=False,      help="(Batch) Wait for the job to finish.")
+    run.add_argument("input",                      metavar="INPUT",              type=str,                     help="Input file to run.")
+    run.add_argument("arguments",                  metavar="ARGUMENTS", nargs="*", type=str, default=[],       help="Additional positional arguments to pass to the case file.")
+    run.add_argument("-e", "--engine",             choices=["interactive", "batch"],              type=str, default="interactive", help="Job execution/submission engine choice.")
```
Comment (sbryngelson) [inline toolchain/mfc/args.py:103]: is there an option to use `-b mpirun` for template files or does one have to modify the template

File: toolchain/mfc/run/run.py [hunk 4]
```diff
@@ -24,30 +31,67 @@ def validate_job_options() -> None:
...
+                       "--cache-control=none", "--clock-control=none"] + ARG("ncu")
 
-    input_file = input.load()
+    if ARG("nsys") is not None:
+        if not does_command_exist("nsys"):
+            raise MFCException("Failed to locate [bold green]NVIDIA Nsight Systems[/bold green] (nsys).")
 
-    engine = engines.get_engine(ARG("engine"))
-    engine.init(input_file)
+        return ["nsys", "profile", "--stats=true", "--trace=mpi,nvtx,openacc"] + ARG("nsys")
```
Comment (sbryngelson) [inline toolchain/mfc/run/run.py:46]: i think i had a comment on your original PR that said that nsys profile needs to go in the opposite order against `mpirun` for whatever we do currently (for mpi jobs at least). does this fix that?

File: toolchain/templates/default.mako [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+warn "Please use a different template via the $MAGENTA--computer$COLOR_RESET option."
+echo
+
+% for binpath in binpaths:
+    echo -e ":) Running $MAGENTA${binpath}$COLOR_RESET:\n"
+
+    % if not mpi:
+        ${' '.join([f"'{x}'" for x in profiler ])} "${binpath}"
+    % else:
+        if command -v jsrun > /dev/null; then
```
Comment (sbryngelson) [inline toolchain/templates/default.mako]: can we not just pass `-b` in here? i know we should be using the templates but this seems like an easy improvement of sorts.

**Review by sbryngelson — CHANGES_REQUESTED** [review]

**Comment by sbryngelson** [comment]
I think this PR was going to fix the issue of not using multiple ranks with `./mfc.sh test -j X` on the Phoenix CPU runner. does it?

File: docs/documentation/running.md (hunk 1 above)
Comment (henryleberre) [reply docs/documentation/running.md]: Fixed! I am getting tired..

File: toolchain/mfc/args.py (hunk 3 above)
Comment (henryleberre) [reply toolchain/mfc/args.py:103]: I agree, I just added it back in.

File: toolchain/templates/default.mako (hunk 5 above)
Comment (henryleberre) [reply toolchain/templates/default.mako]: As per above, I agree and I just added it.

**Comment by henryleberre** [comment]
> I think this PR was going to fix the issue of not using multiple ranks with `./mfc.sh test -j X` on the Phoenix CPU runner. does it?

I had forgotten to include it. Just did. I added:
```console
               --bind-to none                             \
```

So when using GT Phoenix, you just need to use `-c phoenix` while testing and/or running (interactive & batch) MFC.

File: misc/run-phoenix-release-cpu.sh (hunk 2 above)
Comment (henryleberre) [reply misc/run-phoenix-release-cpu.sh]: It forwards arguments to `./mfc.sh run` (called by `./mfc.sh test`). This means we no longer have to redocument all of these options as options of `./mfc.sh test`, too. We also no longer have to manually forward each option.

Comment (henryleberre) [reply misc/run-phoenix-release-cpu.sh]: I just changed this line to:
```console
./mfc.sh test -j $(nproc) -a -- -c phoenix
```

File: toolchain/mfc/run/run.py (hunk 4 above)
Comment (henryleberre) [reply toolchain/mfc/run/run.py:46]: Noted. I just switched the order.

**Comment by sbryngelson** [comment]
so the new command to test on an `mpirun` system is `./mfc.sh test -j X -- --binary mpirun --gpu`? this seems to be what I need when testing your PR on Phoenix.

**Comment by henryleberre** [comment]
> so the new command to test on an `mpirun` system is `./mfc.sh test -j X -- --binary mpirun --gpu`? this seems to be what I need when testing your PR on Phoenix.
> 
> it is also re-building even though the GLOBs are the same?

For Phoenix, I would recommend just:
```console
./mfc.sh test -j X --gpu -- -c phoenix
```

I looked into #288 and it's not quite fixed yet. It is really quite strange. I think it has to do with CMake expecting its binary directory to be empty upon creation skipping GLOB checks within that directory.

**Comment by sbryngelson** [comment]
> > so the new command to test on an `mpirun` system is `./mfc.sh test -j X -- --binary mpirun --gpu`? this seems to be what I need when testing your PR on Phoenix.
> > it is also re-building even though the GLOBs are the same?
> 
> For Phoenix, I would recommend just:
> 
> ```
> ./mfc.sh test -j X --gpu -- -c phoenix
> ```
> 
> But this also works now:
> 
> ```
> ./mfc.sh test -j X --gpu -- -b mpirun -f --bind-to none
> ```
> 
> I looked into #288 and it's not quite fixed yet. It is really quite strange. I think it has to do with CMake expecting its binary directory to be empty upon creation and thus skipping GLOB checks within that directory.

got it

**Comment by henryleberre** [comment]
Part of the appeal of this PR (other than there being a template per system) is that:
- Templates are now used for both Batch & Interactive execution. This gives us more flexibility and makes it a lot easier to understand and modify the behavior of these template files. You no longer need to parse through `toolchain/mfc/run/engines.py`. It doesn't exist anymore!
- Instead of my handcrafted and regex-powered template engine, I switched to an all-around better solution, Mako.
- A lot of indirection was removed (through the support of `--` in argument parsing in many places).
- Refactored `system()` in `toolchain/mfc/` so that whenever we execute a command, we no longer depend on the shell. We provide a list of arguments rather a command-string. It's a lot more difficult to create a bug by introducing spaces in a command or variable name.

**Review by sbryngelson — APPROVED** [review]

**Comment by sbryngelson** [comment]
This looks great - glad to merge once tests pass. We can also now run CPU tests much faster.

**Comment by henryleberre** [comment]
@sbryngelson I just fixed the rebuild issue! Closes #288.

**Comment by sbryngelson** [comment]
> @sbryngelson I just fixed the rebuild issue! Closes #288.

Thank goodness! It was very strange. What was it?

**Comment by sbryngelson** [comment]
@henryleberre you have access to NCSA Delta and Purdue Anvil as well. Can you add appropriate batch templates for those computers?

**Comment by henryleberre** [comment]
> > @sbryngelson I just fixed the rebuild issue! Closes #288.
> 
> Thank goodness! It was very strange. What was it?

There were two issues:
- We only generated the case-optimization FPP files while building, not also while configuring. This causes a GLOB mismatch.
- When I added their generation to the configure step, I had forgotten we also delete the binary directory before running CMake (we have good reasons for doing this). So the files were actually never there while CMake was configuring the MFC targets.

**Comment by henryleberre** [comment]
> @henryleberre you have access to NCSA Delta and Purdue Anvil as well. Can you add appropriate batch templates for those computers?

Could these be added later? It might be a good idea for users to get acquainted with this way of doing things, adding Mako files as they need them.

**Comment by sbryngelson** [comment]
> > @henryleberre you have access to NCSA Delta and Purdue Anvil as well. Can you add appropriate batch templates for those computers?
> 
> Could these be added later? It might be a good idea for users to get acquainted with this way of doing things, adding Mako files as they need them.

Yes that's fine. Merge when appropriate.

**Pull request merged**